include/boost/math/special_functions/jacobi_theta.hpp

```diff
@@ -0,0 +1,646 @@
+
+#ifdef __JACOBI_THETA_USE_IMAGINARY
+// The following _IMAGINARY theta functions assume imaginary z and are for
+// internal use only. The z argument is scaled by tau, and sines and cosines
+// are replaced with hyperbolic sines and cosines to accommodate the imaginary
+// argument. (Recall sinh=(exp(x)-exp(-x))/2 and cosh=(exp(x)+exp(-x))/2)
+//
+// The return values are scaled by exp(-tau*z²/π)/sqrt(tau).
+//
+// Using the convention τ'=-1/τ and noting that both τ and τ' are always
+// imaginary (at least in our applications), the _IMAGINARY functions are used
+// to implement the following four relations:
+//
+// [20.7.30] sqrt(-iτ)θ₁(z|τ) = -i*exp(iτ'z²/π)*θ₁(zτ'|τ')
+// [20.7.31] sqrt(-iτ)θ₂(z|τ) =    exp(iτ'z²/π)*θ₄(zτ'|τ')
+// [20.7.32] sqrt(-iτ)θ₃(z|τ) =    exp(iτ'z²/π)*θ₃(zτ'|τ')
+// [20.7.33] sqrt(-iτ)θ₄(z|τ) =    exp(iτ'z²/π)*θ₂(zτ'|τ')
+template <class RealType, class Policy>
+inline RealType
+_IMAGINARY_jacobi_theta1tau(RealType z, RealType tau, const Policy& pol) {
+    BOOST_MATH_STD_USING
+    unsigned n = 0;
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType /* q_n, */ delta, delta1, delta2, result = RealType(0);
+
+    do {
+        delta1 = exp(tau*(z*(RealType(2*n+1) - z/constants::pi<RealType>())
+                    -constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5)));
+        // q_n = exp(-tau * constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5) );
+        delta2 = exp(tau*(z*(-RealType(2*n+1) - z/constants::pi<RealType>())
+                    -constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5)));
+
+        if (n%2) {
+            delta = delta2 - delta1;
+        } else {
+            delta = delta1 - delta2;
+        }
+
+        result += delta;
+        n++;
+    } while (abs(delta * sqrt(tau)) > eps || (abs(result * sqrt(tau)) > eps * eps && abs(delta/result) > eps));
+    
+    if (abs(result) < eps * eps)
+        return RealType(0);
+
+    return result * sqrt(tau);
+}
+
+template <class RealType, class Policy>
+inline RealType
+_IMAGINARY_jacobi_theta2tau(RealType z, RealType tau, const Policy& pol) {
+    BOOST_MATH_STD_USING
+    unsigned n = 0;
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType /* q_n, */ delta, delta1, delta2, result = RealType(0);
+
+    do {
+        delta1 = exp(tau*(z*(RealType(2*n+1) - z/constants::pi<RealType>())
+                    -constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5)));
+        // q_n = exp(-tau * constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5) );
+        delta2 = exp(tau*(z*(-RealType(2*n+1) - z/constants::pi<RealType>())
+                    -constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5)));
+
+        delta = delta1 + delta2;
+
+        result += delta;
+        n++;
+    } while (abs(delta * sqrt(tau)) > eps || (abs(result * sqrt(tau)) > eps * eps && abs(delta/result) > eps));
+    
+    if (abs(result) < eps * eps)
```

Comment:
I don't get this one; could you add another comment here?

Reply:
I had some tests choking on identities that were supposed to evaluate to zero (e.g. passing in multiples of Pi). I can eliminate this check if you think it's best - it was designed to iron out vanishingly small return values ("epsilon of an epsilon") that the iterator treated as zero anyway.

Reply:
Looks wrong to me as well - very small return values should presumably be possible?

If there's something that should be exactly zero then I'd rather we look at those as special cases.  Also test values that are close to but not exactly zero.

Reply:
If I remove this check, I get some disagreement with Wolfram Alpha (which returns 0 instead of a Very Small Number) for e.g. `jacobi_theta4(0, 0.9921875)`. I will leave the check out for now, and hope that the 1000-bit implementation will clear things up.

Regarding exactly zero, this only happens in a couple of cases, as far as I know: `jacobi_theta1(n*pi, q)` and `jacobi_theta2((n+0.5)*pi, q)`. The special case  `jacobi_theta1(0, q)=0` is already handled, but any other case will slip through the floating-point cracks, unless we re-parameterize z to be a phase (à la `sin_pi`).

Studying the equations for `theta3` and `theta4`, these should never truly return zero. Perhaps a strongly worded letter to Wolfram Alpha is in order on this topic.

Reply:
Agreement with Wolfram Alpha is not a requirement; there are quite a few bugs in Mathematica. Personally I'd rather see unit tests of properties and quadratures to check accuracy.

Reply:
Re: quadrature, the first 3 equations here ought to verify the whole `tau` domain for z=0: https://dlmf.nist.gov/20.10

I previously tried checking relationships with elliptic integrals (e.g. [DLMF 20.9.4](https://dlmf.nist.gov/20.9)), but I ran into some accuracy issues and didn't know which function was to blame. So the existing property checks mostly use simple functions.

Reply:
Added quadrature tests and significantly tightened test tolerances in the latest commit. Quadrature is such an elegant way to verify a function, by the way!

Will revisit the elliptic integral relations tonight.

Reply:
Added tests against Jacobi elliptic functions (`sn`, `cd`, etc) in latest commit, as well as a couple of tests against `ellint_rf`.